src/IO/AsynchronousReadBufferFromFileDescriptor.cpp

```diff
@@ -74,68 +76,43 @@ void AsynchronousReadBufferFromFileDescriptor::prefetch(Priority priority)
 
 bool AsynchronousReadBufferFromFileDescriptor::nextImpl()
 {
+    IAsynchronousReader::Result result;
     if (prefetch_future.valid())
     {
         /// Read request already in flight. Wait for its completion.
 
-        size_t size = 0;
-        size_t offset = 0;
-        {
-            Stopwatch watch;
-            CurrentMetrics::Increment metric_increment{CurrentMetrics::AsynchronousReadWait};
-            auto result = prefetch_future.get();
-            ProfileEvents::increment(ProfileEvents::AsynchronousReadWaitMicroseconds, watch.elapsedMicroseconds());
-            size = result.size;
-            offset = result.offset;
-            assert(offset < size || size == 0);
-        }
+        CurrentMetrics::Increment metric_increment{CurrentMetrics::AsynchronousReadWait};
+        ProfileEventTimeIncrement<Microseconds> watch(ProfileEvents::AsynchronousReadWaitMicroseconds);
 
+        result = prefetch_future.get();
         prefetch_future = {};
-        file_offset_of_buffer_end += size;
-
-        assert(offset <= size);
-        size_t bytes_read = size - offset;
-        if (throttler)
-            throttler->add(bytes_read, ProfileEvents::LocalReadThrottlerBytes, ProfileEvents::LocalReadThrottlerSleepMicroseconds);
-
-        if (bytes_read)
-        {
+        if (result.size - result.offset > 0)
             prefetch_buffer.swap(memory);
-            /// Adjust the working buffer so that it ignores `offset` bytes.
-            internal_buffer = Buffer(memory.data(), memory.data() + memory.size());
-            working_buffer = Buffer(memory.data() + offset, memory.data() + size);
-            pos = working_buffer.begin();
-            return true;
-        }
-
-        return false;
     }
     else
     {
         /// No pending request. Do synchronous read.
 
-        Stopwatch watch;
-        auto [size, offset, _] = asyncReadInto(memory.data(), memory.size(), DEFAULT_PREFETCH_PRIORITY).get();
-        ProfileEvents::increment(ProfileEvents::AsynchronousReadWaitMicroseconds, watch.elapsedMicroseconds());
-
-        file_offset_of_buffer_end += size;
+        ProfileEventTimeIncrement<Microseconds> watch(ProfileEvents::SynchronousReadWaitMicroseconds);
+        result = asyncReadInto(memory.data(), memory.size(), DEFAULT_PREFETCH_PRIORITY).get();
+    }
 
-        assert(offset <= size);
-        size_t bytes_read = size - offset;
-        if (throttler)
-            throttler->add(bytes_read, ProfileEvents::LocalReadThrottlerBytes, ProfileEvents::LocalReadThrottlerSleepMicroseconds);
+    chassert(result.size >= result.offset);
+    size_t bytes_read = result.size - result.offset;
+    file_offset_of_buffer_end += result.size;
```

Comment:
Not related to this PR, but shouldn't this be `result.size - result.offset`?

Reply:
`result.offset` is the same as `bytes_to_ignore` here actually and `bytes_to_ignore < internal_buffer.size() && bytes_to_ignore (== result.offset) + result.size <= internal_buffer.size()` (there is an assert for this in `seek()`). And as `bytes_to_ignore` was not accounted in `file_offset_of_buffer_end` yet, we need to account it here.

Reply:
> result.offset is the same as bytes_to_ignore here actually

Why? Does IAsynchronousReader pass-through `result.offset = request.ignore`? I guess it's not true for all IAsynchronousReader implementations, but only the one(s) that can be used with AsynchronousReadBufferFromFileDescriptor? Consider adding a comment about it. (Pretty likely that I'm misreading the code, and these questions are nonsense - feel free to ignore then, maybe still add some comment if that's easy to do.)